File: backend/src/utils/haversine.ts
const EARTH_RADIUS_METERS = 6_371_000;

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function haversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_METERS * c;
}

export interface ProximityResult {
  distance: number;
  withinRange: boolean;
  isRemote: boolean;
}

export function checkProximity(
  checkInLat: number,
  checkInLng: number,
  locationLat: number,
  locationLng: number,
  maxDistance = 200,
  remoteThreshold = 100,
): ProximityResult {
  const distance = haversineDistance(checkInLat, checkInLng, locationLat, locationLng);
  return {
    distance: Math.round(distance),
    withinRange: distance <= maxDistance,
    isRemote: distance > remoteThreshold,
  };
}

File: app/lib/core/location/haversine.dart
import 'dart:math';

/// Pure-Dart implementation of the Haversine formula.
///
/// No external dependencies -- uses only `dart:math`.
class Haversine {
  Haversine._();

  /// Mean radius of the Earth in metres (WGS-84).
  static const double _earthRadiusMeters = 6371008.8;

  /// Returns the great-circle distance in **metres** between two points
  /// specified in decimal degrees.
  static double distanceInMeters(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    final dLat = _toRadians(lat2 - lat1);
    final dLon = _toRadians(lon2 - lon1);

    final a = sin(dLat / 2) * sin(dLat / 2) +
        cos(_toRadians(lat1)) *
            cos(_toRadians(lat2)) *
            sin(dLon / 2) *
            sin(dLon / 2);

    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return _earthRadiusMeters * c;
  }

  /// Convenience: returns distance in **kilometres**.
  static double distanceInKm(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    return distanceInMeters(lat1, lon1, lat2, lon2) / 1000.0;
  }

  static double _toRadians(double degrees) => degrees * pi / 180.0;
}
